timestamp/src/sorting.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::account::AccountManager;
+pub use ckey::Ed25519Public as Public;
+use coordinator::module::*;
+use coordinator::{Transaction, TransactionWithMetadata};
+use foundry_module_rt::UserModule;
+use parking_lot::RwLock;
+use remote_trait_object::raw_exchange::{import_service_from_handle, HandleToExchange, Skeleton};
+use remote_trait_object::{service, Context as RtoContext, Service};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[service]
+pub trait GetAccountAndSeq: Service {
+    fn get_account_and_seq(&self, tx: &Transaction) -> Result<(Public, u64), ()>;
+}
+
+struct Context {
+    account_manager: Option<Box<dyn AccountManager>>,
+    get_account_and_seqs: HashMap<String, Box<dyn GetAccountAndSeq>>,
+}
+
+impl Service for Context {}
+
+impl Context {
+    fn account_manager(&self) -> &dyn AccountManager {
+        self.account_manager.as_ref().unwrap().as_ref()
+    }
+}
+
+impl TxSorter for Context {
+    // TODO: Consider origin
+    fn sort_txs(&self, txs: &[TransactionWithMetadata]) -> SortedTxs {
+        // TODO: Avoid Public hashmap
+        let mut accounts: HashMap<Public, Vec<(u64, usize)>> = HashMap::new();
+        let mut invalid: Vec<usize> = Vec::new();
+
+        for (i, tx) in txs.iter().enumerate() {
+            if let Some(get_account_and_seq) = self.get_account_and_seqs.get(tx.tx.tx_type()) {
+                if let Ok((public, seq)) = get_account_and_seq.get_account_and_seq(&tx.tx) {
+                    if let Some(list) = accounts.get_mut(&public) {
+                        list.push((seq, i));
+                    } else {
+                        accounts.insert(public, vec![(seq, i)]);
+                    }
+                } else {
+                    invalid.push(i);
+                }
+            } else {
+                invalid.push(i);
+            }
+        }
```

Comment:
Let's split this loop into two parts.
* Getting (public, account)
* Using (public, account)

For example:
```rust
 impl Context {
     fn account_and_seq_from_tx(&self, tx: &TransactionWithMetadata) -> Option<(Public, u64)> {
         let get_account_and_seq: &Box<dyn GetAccountAndSeq> = match self.get_account_and_seqs.get(tx.tx.tx_type()) {
             Some(get_account_and_seq: &Box<dyn GetAccountAndSeq>) => get_account_and_seq,
             None => return None,
         };

         match get_account_and_seq.get_account_and_seq(&tx.tx) {
             Ok((public: Public, seq: u64)) => return Some((public, seq)),
             _ => return None,
         }
     }
 }


 impl TxSorter for Context {
     // TODO: Consider origin
     fn sort_txs(&self, txs: &[TransactionWithMetadata]) -> SortedTxs {
         // TODO: Avoid Public hashmap
         let mut accounts: HashMap<Public, Vec<(u64, usize)>> = HashMap::new();
         let mut invalid: Vec<usize> = Vec::new();

          for (i: usize, tx: &TransactionWithMetadata) in txs.iter().enumerate() {
             if let Some((public: Public, seq: u64)) = self.account_and_seq_from_tx(tx) {
                 if let Some(list: &mut Vec<(u64, usize)>) = accounts.get_mut(&public) {
                     list.push((seq, i));
                 } else {
                     accounts.insert(k: public, vec![(seq, i)]);
                 }
             } else {
                 invalid.push(i);
             }
         }


```